README.md
eea.plonebuildout.profile
=========================

EEA Plone buildout profile


docs/HISTORY.txt
Changelog
=========

4.8-dev0 - (unreleased)
---------------------------

4.7 - (2021-03-29)
---------------------------
* Change: Add dependency to eea.geolocation
  [avoinea]

4.6 - (2020-04-01)
--------------------------
* Feature: implemented Manifest.json renderer and editing tools.
  [iugin refs #115143]

4.5 - (2019-05-16)
--------------------------
* Change: Moved raven processor to Sentry dedicated package eea.sentry
  [avoinea refs #104872]

4.4 - (2019-01-28)
--------------------------
* Jenkins: Add sonarqube step
  [avoinea refs #101552]

4.3 - (2018-09-25)
--------------------------
* Cleanup: Remove eea.controlpanel dependency
  [avoinea refs #95687]

4.2 - (2018-09-12)
--------------------------
* Upgrade step: Within "Plone > Site setup > Add-ons" click on
  upgrade button available for EEA Plone buildout profile.
  [avoinea refs #95172]
* Change: Include jquery-migrate plugin
  [avoinea refs #95172]

4.1 - (2018-09-10)
--------------------------
* Cleanup: Remove eea.socialmedia dependency
  [avoinea refs #95341]

4.0 - (2018-07-30)
--------------------------
* Feature: added 'cssselect' as dependency, required by lxml version 3.x
  [alecghica refs #96974]

3.9 - (2018-06-20)
--------------------------
* Change: updated URLs pointing to eea.europa.eu with https://
  [alecghica refs #95849]

3.8 - (2018-06-12)
--------------------------
* Feature: Add custom raven processors to cleanup Zope REQUEST
  [avoinea refs #95646]

3.7 - (2018-06-12)
------------------
* Change: show new kgs version as warning instead of error in order to avoid
  ajax loading of the error instead of having a page redirect on login popup
  form
  [ichim-david refs #92454]

3.6 - (2018-03-01)
------------------
* Feature: Added dependency on ftw.copymovepatches in order to speedup Cut/Copy
  operations
  [avoinea refs #92869]

3.5 - (2018-02-02)
------------------
* Feature: Added dependency on ftw.globalstatusmessage
  [avoinea refs #92245]

3.4 - (2017-12-22)
------------------
* Change: Update latest release viewlet to check over eea.docker.kgs
  [avoinea refs #89878]

3.3 - (2017-12-12)
------------------
* Change: Replace eeacms/zptlint with eeacms/plone-test:4 zptlint
  [avoinea refs #90415]

3.2 - (2017-04-24)
------------------
* Change: updated package information
  [eduard-fironda]

3.1 - (2017-04-12)
------------------
* Feature: enabled eea.similarity as a default add-on for EEA CPB
  [alecghica refs #82296]

3.0 - (2017-03-27)
------------------
* Bug fix: Fixed  JS error `TypeError: _widget` when eea.indicators is installed
  [avoinea refs #83106]
* Change: Do not load eea.relations.default SchemaModifier if `eea.design` is
  installed
  [avoinea refs #83106]
* Change: Remove `[full]` dependency on eea.daviz
  [avoinea refs #83106]
* Change: Remove dependency to Products.kupu
  [avoinea]

2.8 - (2017-01-17)
------------------
* Change: Add dependency to eea.async.manager
  [voineali refs #80206]

2.7 - (2016-07-07)
------------------
* Change: Pin eea.facetednavigation < 10.0
  [avoinea]

2.6 - (2016-06-02)
------------------
* Change: Add dependency to collective.warmup
  [voineali refs #72395]

2.5 - (2016-02-16)
------------------
* Bug fix: Don't show any viewlet if this package is not installed
  [voineali refs #68842]

2.4 - (2015-09-17)
------------------
* Bug fix: Add dependencies to setup.py and update tests
  [voineali refs #27405]

2.3 - (2015-07-21)
------------------
* Bug fix: get_hostname returns a list and was not propely handled
  [larremik refs #26682]

2.2 - (2015-06-02)
------------------
* Bug fix: Release a new egg
  [ghicaale refs #25014]

2.1 - (2015-05-27)
------------------
* Bug fix: fixed hostname filter to proper display CPB analytics
  [ghicaale]

2.0 - (2015-05-19)
------------------
* Upgrade step: Within "Plone > Site setup > Add-ons" click on
  upgrade button available for EEA Plone buildout profile.
  [voineali refs #24289]
* Feature: Use **plone.app.ldap** to manage LDAP and Active Directory servers
  [voineali refs #24289]

1.1 - (2015-02-16)
------------------
* Bug fix: filter bad/irrelevant domain names
  [ghicaale refs #22687]

1.0 - (2015-02-03)
------------------
* Feature: Add viewlet with required/mandatory packages
  [olimpiurob refs #22038]

0.9 - (2015-01-08)
------------------
* Upgrade step: Within "Plone > Site setup > Add-ons" click on
  upgrade button available for EEA Plone buildout profile
  [olimpiurob refs #22037]
* Change: Added IPloneBuildoutProfileLayer add-on layer
  [olimpiurob refs #22037]

0.8 - (2015-01-08)
------------------
* Feature: Added analytics reporting feature
  [olimpiurob refs #22039]

0.7 - (2014-11-20)
------------------
* Feature: added gitignore list
  [ghicaale]

0.6 - (2014-02-03)
------------------
* Feature: Add 'is_view_template' condition to KGS availability viewlet template
  [olimpiurob #17960]
* Bug fix: Fixed an issue where logging the github api response returned
  an error which stated that you can't mix a string and a response
  [olimpiurob #17960]

0.5 - (2013-12-16)
------------------
* Feature: Manager is announced of any buildout changes, not just kgs
  [simiamih refs #17343]
* Bugfix: LDAP_INSTALLED NameError for new installations
  [simiamih refs #17343]

0.4 - (2013-11-05)
------------------
* Change: cache the KGS version for 24 hours
  [tiberich]

0.3 - (07-10-2013)
----------------------
* Bug fix: treat case where github does not return proper json response
  [tiberich]

0.2 - (11-07-2013)
----------------------
* Feature: added a viewlet that informs managers of new EEA KGS availability
  [tiberich #14827]
* Change: fixed profile name and cleanup empty spaces.
  [ghicaale refs #14827]
* Bug fix: don't setup EIONET LDAP if LDAP support is not available.
  [ghicaale refs #14827]

0.1 (04-07-2013)
--------------------
* Initial release. A simple GS profile that depends on all the core EEA
  packages and sets up the LDAP based user authentication.
  [tiberich #14827]


eea/plonebuildout/profile/browser/viewlets.py
"""Viewlets for eea.plonebuildout.profile
"""
import os
import logging
import requests
import contextlib
from time import time

from Products.CMFCore.utils import getToolByName
from distutils.version import StrictVersion
from eea.plonebuildout.profile.browser.utils import REQUIRED_PKGS
from plone.app.layout.viewlets.common import ViewletBase
from plone.memoize import ram

from zope.component import getUtility
from eea.plonebuildout.profile.controlpanel.manifest_json import IManifestJsonSettings
from plone import api
from plone.registry.interfaces import IRegistry

logger = logging.getLogger('eea.plonebuildout.profile')
EEA_KGS_URL = "https://api.github.com/repos/eea/eea.docker.kgs/tags"
READ_TIMEOUT = 3.0
CONNECT_TIMEOUT = 3.0


class NewReleaseViewlet(ViewletBase):
    """ A viewlet which informs managers of new EEA KGS releases
    """

    def current_kgs_version(self):
        """ Get current KGS version running on
        """
        return os.environ.get("EEA_KGS_VERSION", '')

    #we cache the result for 24 hours
    @ram.cache(lambda *args:time() // (60*60*24))
    def last_kgs_update(self):
        """ Return a version number if running old KGS version, otherwise None
        """

        kgsver = self.current_kgs_version()
        if not kgsver:
            logger.warn("EEA_KGS_VERSION is not defined as environment "
                        "variable. Please use proper buildout configuration")
            return ''

        with contextlib.closing(requests.get(EEA_KGS_URL)) as conn:
            tags = conn.json()
            if not (tags and isinstance(tags, list)):
                logger.warn("Could not retrive KGS version at %s", EEA_KGS_URL)
                return ''

            last = tags[0].get('name', None)
            try:
                if StrictVersion(last) > StrictVersion(kgsver):
                    return last
            except (ValueError, AttributeError):
                if last != kgsver:
                    return last
        return ''


class RequiredPkgsViewlet(ViewletBase):
    """ A viewlet which displays a warning about missing required packages
    """

    def get_missing_packages(self):
        """ Return a list of missing required packages
        """
        missing = {}
        available_pkgs = []
        installed_pkgs = []
        qi = getToolByName(self.context, 'portal_quickinstaller')
        prods = qi.listInstallableProducts(skipInstalled=False)

        for prod in prods:
            pkg_id = prod.get('id')
            if prod.get('status') == 'installed':
                installed_pkgs.append(pkg_id)
            else:
                available_pkgs.append(pkg_id)

        for pkg in REQUIRED_PKGS:
            pkg_id = pkg.get('pkg_id')
            if pkg_id not in installed_pkgs:
                if pkg_id not in available_pkgs:
                    try:
                        __import__(pkg_id)
                    except ImportError:
                        missing_pkg = pkg.copy()
                        missing_pkg['available'] = False
                        missing[pkg_id] = missing_pkg
                else:
                        missing_pkg = pkg.copy()
                        missing_pkg['available'] = True
                        missing[pkg_id] = missing_pkg

        return missing

class ManifestJsonViewlet(ViewletBase):
    def get_manifest_json_settings(self):
        """ Return the settings as set in site/@@pwa-controlpanel

            Usage: s.name
        """
        registry = getUtility(IRegistry, context=api.portal.get())
        try:
            s = registry.forInterface(IManifestJsonSettings)
        except KeyError:
            s = object()
        return s

    def get_theme_color(self):
        settings = self.get_manifest_json_settings()
        if hasattr(settings, 'theme_color'):
            return settings.theme_color

        return ''
